### PR TITLE
Cherrypick fixes for #16368 into Bazel 6.3.0

### PR DESCRIPTION
**Pull request by ahumesky**
Cherrypick the following changes into 6.3.0 to address #16368 

c89ca0d4d772b08fd852bcd35efd077253b02cc9
166ef61844d35f63944962011a12de794334522d
fbb58430a7b55f7e8a9f6175d96a93772816b007
8d2ccd47569707155410aac035bd8a7a8d126c6d
9092303c358b87c1e027abafe135c44a27531b36

### CODE REVIEW

**Comment by ahumesky** [comment]
The CLA bot is reporting that the CLA is missing for `Author: @a-googler <no****y​@google.com>`, which is surprising...

**Comment by keertk** [comment]
Thanks @ahumesky. I think @meteorcloudy can help force merge this PR once approved.

**Review by meteorcloudy — APPROVED** [review]

**Comment by ahumesky** [comment]
Thank you!

**Pull request merged**
